#162725256 Users should be able to reset passwords via email

malfahad opened the pull request:
#### What does this PR do?
Enable users to reset their password via email. 


#### Description of Task to be completed?
This PR creates the following endpoints
`POST /api/users/password-reset`
 - Accepts a JSON body with an email field.
 - Email field is matched to an existing account
 - A password reset link is generated
 - The 'password reset email' is sent to the registered email address.


`GET /api/users/password-reset/<token>`
 - Reads and validates the password reset token

`POST /api/users/password-reset/<token>`
 - Accepts a password field with a new password
- updates the user with a new password
 
#### How should this be manually tested?

 1. Fetch and checkout this branch with `git fetch origin ft-reset-password-162725256 && git checkout ft-reset-password-162725256` 
2. activte your virtual environeent
3. Install dependencies with `pip install -r requirements.txt`
4. add the following to your `env` file.
    ```
    EMAIL_HOST_USER = <email address goes here>
    EMAIL_HOST_PASSWORD = <email password goes here>
   ```
5. Run tests with `python manage.py test`
6. Run the server `python manage.py runserver`


#### How do i test the endpoints?
#### First endpoint:

 In postman, make a POST request to `/api/users/password-reset/` with:
```  
{
	"user":{
	"email":"sample.email@host.com"
    }
}
```

 where `sample.email@host.com` is a registered email in your database. An email with a recent link will be sent.

#### Second Endpoint
 In postman, make a POST request to `/url/sent/in/email` with:
```
{
	"user":{
		"password":"new.password"
	}
}
```

The registered password will be updated. To verify, [login with the new credentials](https://github.com/andela/Ah-backend-xmen#authentication).

#### What are the relevant pivotal tracker stories?
Pivotal Tracker Story: [#162725256](https://www.pivotaltracker.com/story/show/162725256)

#### Postman documentation
[![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/63c458e450e059619049#?env%5Bauthors-local%5D=W3sia2V5IjoiYmFzZS11cmwiLCJ2YWx1ZSI6IiIsImRlc2NyaXB0aW9uIjoiIiwiZW5hYmxlZCI6dHJ1ZX0seyJrZXkiOiJyZXNldC10b2tlbiIsInZhbHVlIjoiIiwiZGVzY3JpcHRpb24iOiIiLCJ0eXBlIjoidGV4dCIsImVuYWJsZWQiOnRydWV9XQ==)

MuhweziDeo commented on authors/settings.py:
```diff
@@ -30,13 +33,24 @@
+EMAIL_USE_TLS = True
+EMAIL_PORT = 587
+EMAIL_HOST_USER = config('EMAIL_HOST_USER')
+EMAIL_HOST_PASSWORD = config('EMAIL_HOST_PASSWORD')
```
set these keys in travis

MuhweziDeo commented:
@malfahad  it appears u didn't rebase to develop after @armstrongsouljah PR was merged

MuhweziDeo reviewed (dismissed):
Please first rebase

Commit:
feat(password-reset): reset password via email

- created core model class to handle password reset
- configured smtp settings with gmail to send sms
- added url endpoints for password reset requests and password reset
- wrote test for password reset request

[STARTS #162725256]

ja-odur commented:
@malfahad Please do not include sensitive information such as passwords in the `PR`. simply update the sample env file and whoever wants to test the `PR` can get the credentials from you or you share the credentials on the slack channel.

ja-odur commented on .travis.yml:
```diff
@@ -5,7 +5,7 @@ python:
     - postgres
 
-# require travis to first install dependencies
+   # require travis to first install dependencies
```
@malfahad Please undo this change.

malfahad replied:
Thanks for the heads-up. I have made those changes here [548a69b](https://github.com/andela/Ah-backend-xmen/commit/548a69bde95c9c63f97d6c94bdb14b9ffa9e73ff)

malfahad commented:
@ja-odur I updated the PR without the sensitive info here [7db95bf](https://github.com/andela/Ah-backend-xmen/commit/7db95bf36419dcfec32572747e2bb7399a0b9afa)

ja-odur commented:
> 7db95bf

@malfahad in the section `how it can be manually tested` bullet four, aren't those the actual email credentials.

malfahad commented:
> > [7db95bf](https://github.com/andela/Ah-backend-xmen/commit/7db95bf36419dcfec32572747e2bb7399a0b9afa)
> 
> @malfahad in the section `how it can be manually tested` bullet four, aren't those the actual email credentials.

I had completely missed that and only changed the travis.yml file. But have made the changes now.

MuhweziDeo reviewed (dismissed):
I like the rendering email message to HTML. the message looks good 
your good to go

ja-odur commented on authors/apps/core/models.py:
```diff
@@ -0,0 +1,78 @@
+
+        #generate user token to be used in password reset link
+        token = jwt.encode({'email': user.email,
+                             'exp': datetime.datetime.utcnow() + datetime.timedelta(days=1)
```
@malfahad please set the expiry time to 30 minutes instead of 1 hour.

ja-odur commented on authors/apps/authentication/views.py:
```diff
@@ -72,3 +74,52 @@ def update(self, request, *args, **kwargs):
+        serializer.is_valid(raise_exception=True)
+        user.set_password(serializer.data.get('password'))
+        user.save()
+        return Response({"password-reset":"Your password has been updated"} , status=status.HTTP_200_OK)
```
@malfahad add a new line at the end of this file.

ja-odur reviewed (approved):
LGTM

kdaglas reviewed (approved):
I like the messages, and the referencing, they are precise and straight to the point, LGTM

armstrongsouljah reviewed (approved):
LGTM, good work on the validations too

d-rita reviewed (commented):
works fine.
